src/libraries/System.Diagnostics.DiagnosticSource/src/System/Diagnostics/W3CPropagator.cs

```diff
@@ -0,0 +1,741 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Runtime.CompilerServices;
+using System.Text;
+
+namespace System.Diagnostics
+{
+    internal sealed class W3CPropagator : DistributedContextPropagator
+    {
+        internal static DistributedContextPropagator Instance { get; } = new W3CPropagator();
+
+        private const int MaxBaggageEntriesToEmit = 64;     // Suggested by W3C specs
+        private const int MaxBaggageEncodedLength = 8192;   // Suggested by W3C specs
+        private const int MaxTraceStateEncodedLength = 256; // Suggested by W3C specs
+
+        private const char Equal = '=';
+        private const char Percent = '%';
+        private const char Replacement = '\uFFFD'; // �
+
+        public override IReadOnlyCollection<string> Fields { get; } = new ReadOnlyCollection<string>(new[] { TraceParent, TraceState, Baggage });
+
+        public override void Inject(Activity? activity, object? carrier, PropagatorSetterCallback? setter)
+        {
+            if (activity is null || setter is null || activity.IdFormat != ActivityIdFormat.W3C)
+            {
+                return;
+            }
+
+            string? id = activity.Id;
+            if (id is null)
+            {
+                return;
+            }
+
+            setter(carrier, TraceParent, id);
+            if (!string.IsNullOrEmpty(activity.TraceStateString))
+            {
+                InjectTraceState(activity.TraceStateString, carrier, setter);
+            }
+
+            InjectW3CBaggage(carrier, activity.Baggage, setter);
+        }
+
+        public override void ExtractTraceIdAndState(object? carrier, PropagatorGetterCallback? getter, out string? traceId, out string? traceState)
```

Comment:
nit: for better readability and clarity:

```suggestion
        public override void ExtractTraceIdAndState(object? carrier, PropagatorGetterCallback? getter, out string? traceParent, out string? traceState)
```

Reply:
Unfortunetly, this has been [exposed](https://learn.microsoft.com/en-us/dotnet/api/system.diagnostics.distributedcontextpropagator.extracttraceidandstate?view=net-9.0) with this name. Changing it now is kind of a breaking change. Renaming parameter names needs to get reviewed and approved by the design committee. Please let me know if you think it is important enough to rename and I can track it.

Reply:
Lets correct it - changing the name won't break code bindings as the type is the same.

Reply:
yeah, I mean just the local param name, I don't believe method rename is worth it

Reply:
This is not true. It is a breaking change. If we need to change it, we need to design review it. I am not seeing it is worth changing it. 

---
https://learn.microsoft.com/en-us/dotnet/core/compatibility/library-change-rules

❌ DISALLOWED: Renaming a parameter (including changing its case)

This is considered breaking for two reasons:

It breaks late-bound scenarios such as the late binding feature in Visual Basic and [dynamic](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/reference-types#the-dynamic-type) in C#.

It breaks [source compatibility](https://learn.microsoft.com/en-us/dotnet/core/compatibility/categories#source-compatibility) when developers use [named arguments](https://learn.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/named-and-optional-arguments#named-arguments).